src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,445 @@
+
+(*** Lemmas on prefixes ***)
+/// TODO: These may belong in DY.Core.Trace.Base alongside the other prefix lemmas.
+
+let nil_prefix (#label_t:Type) (tr:trace_ label_t)
+  : Lemma (ensures Nil <$ tr)
+    [SMTPat (Nil <$ tr)]
+  = normalize_term_spec (grows #label_t)
+
+let grows_snoc_right (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures tr1 <$ (Snoc tr2 e))
+    [SMTPat (tr1 <$ (Snoc tr2 e)); SMTPat (tr1 <$ tr2)]
+  = normalize_term_spec (grows #label_t)
+
+let grows_snoc_left (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (requires (Snoc tr1 e) <$ tr2)
+    (ensures tr1 <$ tr2)
+    [SMTPat (tr1 <$ tr2); SMTPat ((Snoc tr1 e) <$ tr2)]
+  = grows_snoc_right tr1 tr1 e
+
+let prefix_full_eq (#label_t:Type) (tr:trace_ label_t)
+  : Lemma (ensures (prefix tr (length tr) == tr))
+    [SMTPat (prefix tr (length tr))]
+  = normalize_term_spec (grows #label_t)
+
+let grows_full_eq (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2 /\ length tr1 == length tr2)
+    (ensures tr1 == tr2)
+// TODO:
+// It is not clear if this SMTPat is ever practically useful, but it seems like it could be in principle
+    [SMTPat (tr1 == tr2); SMTPat (tr1 <$ tr2)]
+  = prefix_prefix_eq tr1 tr2 (length tr1)
+
+let grows_cases (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures tr1 == tr2 \/ (
+      Snoc? tr2 /\ (
+      let Snoc hd _ = tr2 in
+      tr1 <$ hd
+    )))
+  = if length tr1 = length tr2
+    then grows_full_eq tr1 tr2
+    else begin
+      let Snoc hd e = tr2 in
+      calc (<$) {
+        tr1;
+        == {}
+        prefix tr1 (length tr1);
+        <$ { prefix_prefix_grows tr1 tr2 (length tr1) (length hd) }
+        prefix tr2 (length hd);
+        == { prefix_prefix_eq hd tr2 (length hd) }
+        prefix hd (length hd);
+        == { }
+        hd;
+      }
+    end
+
+
+(*** General utility functions for working with traces ***)
+
+// This function is not terribly important, but we often want to check if we are at
+// the end of a trace, and hiding the off-by-one between length and the last index
+// inside a function makes things look cleaner.
+let last_ts (#label_t:Type) (tr:trace_ label_t{Snoc? tr}) : timestamp = length tr - 1
+
+// Now that trace entries are abstracted over label types, we cannot compare
+// trace entries for equality with =. This function implements trace equality
+// when the label type is an eqtype (as for unit, for instance).
+let trace_entry_equal (#label_t:eqtype) (e1 e2:trace_entry_ label_t)
+  : bool
+  = match (e1, e2) with
+    | MsgSent b1, MsgSent b2 -> b1 = b2
+    | RandGen u1 l1 i1, RandGen u2 l2 i2 -> u1 = u2 && l1 = l2 && i1 = i2
+    | Corrupt ts1, Corrupt ts2 -> ts1 = ts2
+    | SetState p1 sid1 b1, SetState p2 sid2 b2 -> p1 = p2 && sid1 = sid2 && b1 = b2
+    | Event p1 t1 b1, Event p2 t2 b2 -> p1 = p2 && t1 = t2 && b1 = b2
+    | _ -> false
```

Comment:
I have been testing, we can remove the `noeq` from the definition of `trace_entry_` and `trace_`, in that case when `label_t` is an `eqtype`, then `trace_entry_` is also an `eqtype`, hence we would be able to define `trace_entry_equal e1 e2 = (e1 = e2)`. This means we don't really need this function, nor would we need the lemma `trace_entry_equal_is_eq`